sympy/core/power.py

```diff
@@ -407,7 +407,14 @@ def _eval_Mod(self, q):
             exponentiation.
             '''
             if self.base.is_Integer and self.exp.is_Integer and q.is_Integer:
-                return pow(int(self.base), int(self.exp), int(q))
+                b, e, m = int(self.base), int(self.exp), int(q)
+                # for very large powers
+                if m<=10**18 and _log(e,2)>4*m**(1/4): # pollard rho for modulo would not be slower than binary exponentiation
+                    from sympy.ntheory import totient
+                    phi = totient(m)
+                    return pow(b,phi + e%phi,m) # reduction by euler totient theorem and crt
```

Comment:
Is the extra `phi` essential?

Reply:
Yes it is essential, conventional euler totient theorem holds only for coprime base and modulo (a^n mod m = a^(n mod phi(m)) mod m, ie, when gcd(a,m)=1).